upb/decode_fast.c

```diff
@@ -140,77 +143,31 @@ UPB_FORCEINLINE uint64_t fastdecode_munge(uint64_t val, int valbytes, bool zigza
   return val;
 }
 
-UPB_FORCEINLINE
-static int fastdecode_varintlen(uint64_t data64) {
-  uint64_t clear_bits = ~data64 & 0x8080808080808080;
-  if (clear_bits == 0) return -1;
-  return __builtin_ctzl(clear_bits) / 8 + 1;
-}
-
-UPB_FORCEINLINE
-static const char *fastdecode_longvarint(UPB_PARSE_PARAMS, int valbytes,
-                                         int varintbytes, bool zigzag) {
-  uint64_t val = data >> 18;
-  size_t ofs = (uint16_t)data;
-  uint64_t data64;
-  int sawbytes;
-  memcpy(&data64, ptr + 2, 8);
-  sawbytes = fastdecode_varintlen(data64) + 2;
-  UPB_ASSERT(sawbytes == varintbytes);
-#ifdef __BMI2__
-  if (varintbytes != 3) {
-    uint64_t mask = 0x7f7f7f7f7f7f7f7f >> (8 * (10  - varintbytes));
-    val |= _pext_u64(data64, mask) << 14;
-  } else
-#endif
-  {
-    int i;
-    for (i = 2; i < varintbytes; i++) {
-      uint64_t byte = ptr[i];
-      if (i != varintbytes - 1) byte &= 0x7f;
-      val |= byte << (7 * i);
-    }
-  }
-  val = fastdecode_munge(val, valbytes, zigzag);
-  memcpy((char*)msg + ofs, &val, valbytes);
-  return fastdecode_dispatch(d, ptr + varintbytes, msg, table, hasbits);
-}
-
-UPB_FORCEINLINE
-static const char *fastdecode_longvarintjmp(UPB_PARSE_PARAMS,
-                                            _upb_field_parser **funcs) {
-  int len;
-  uint64_t data64;
-  memcpy(&data64, ptr + 2, 8);
-  len = fastdecode_varintlen(data64);
-  if (len < 0) return fastdecode_err(d);
-  return funcs[len - 1](UPB_PARSE_ARGS);
-}
-
 UPB_FORCEINLINE
 static const char *fastdecode_varint(UPB_PARSE_PARAMS, int tagbytes,
-                                     int valbytes, upb_card card, bool zigzag,
-                                     _upb_field_parser **funcs) {
+                                     int valbytes, upb_card card, bool zigzag) {
   uint64_t val;
   void *dst;
   if (UPB_UNLIKELY(!fastdecode_checktag(data, tagbytes))) {
     RETURN_GENERIC("varint field tag mismatch\n");
   }
-  dst = fastdecode_getfield(d, ptr, msg, &data, &hasbits, tagbytes, valbytes,
+  dst = fastdecode_getfield(d, ptr, msg, &data, &hasbits, valbytes,
                             card);
-  val = (uint8_t)ptr[tagbytes];
+  ptr += tagbytes + 1;
+  val = (uint8_t)ptr[-1];
   if (UPB_UNLIKELY(val & 0x80)) {
-    uint32_t byte = (uint8_t)ptr[tagbytes + 1];
-    val += (byte - 1) << 7;
-    if (UPB_UNLIKELY(byte & 0x80)) {
-      ptr += tagbytes;
-      data = (uint32_t)(val << 18 | data >> 48);
-      return fastdecode_longvarintjmp(UPB_PARSE_ARGS, funcs);
+    for (int i = 0; i < 8; i++) {
```

Comment:
This has pros and cons:

+ solves the "push rbx" problem for short varints
+ can handle repeated varints

- possibly less efficient for longer varints
- larger code size

On balance I think it's a good move for now.

Reply:
I agree, this is just a for now thing. Ideally we find a branch free BMI version here.